pkg/trace/stats/concentrator.go

```diff
@@ -216,6 +237,69 @@ func (c *Concentrator) Add(t Input) {
 	c.mu.Unlock()
 }
 
+// ProcessOTLPTraces applies APM stats calculation on the otlp traces in the concentrator.
+// This function is NOT called in Concentrator.Run(), so you need to manually call it to get APM stats on the given OTLP traces.
+func (c *Concentrator) ProcessOTLPTraces(traces ptrace.Traces) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	includeCID := shouldIncludeCIDDim(c.conf)
+	spanByID, resByID, scopeByID := IndexOTelSpans(traces)
+	topLevelByKind := !c.conf.HasFeature("disable_otlp_compute_top_level_by_span_kind")
+	topLevelSpans := GetTopLevelOTelSpans(spanByID, resByID, topLevelByKind)
+	ignoreResNames := make(map[string]struct{})
+	for _, resName := range c.conf.Ignore["resource"] {
+		ignoreResNames[resName] = struct{}{}
+	}
+	for spanID, span := range spanByID {
+		res := resByID[spanID]
+		resName := GetOTelResource(span, res)
+		if _, exists := ignoreResNames[resName]; exists {
+			continue
+		}
+		env := GetOTelAttrValInResAndSpanAttrs(span, res, doNormalize, semconv.AttributeDeploymentEnvironment)
+		if env == "" {
+			env = c.agentEnv
+		}
+		aggKey := PayloadAggregationKey{
+			Env:      env,
+			Hostname: GetOTelHostname(span, res, c.conf),
+			Version:  GetOTelAttrValInResAndSpanAttrs(span, res, doNormalize, semconv.AttributeServiceVersion),
+		}
+		if includeCID {
+			if cid := GetOTelAttrValInResAndSpanAttrs(span, res, doNormalize, semconv.AttributeContainerID, semconv.AttributeK8SPodUID); cid != "" {
+				aggKey.ContainerID = cid
+				if gitCommitSha, imageTag, err := version.GetVersionDataFromContainerTags(cid, c.conf); err == nil {
+					aggKey.GitCommitSha = gitCommitSha
+					aggKey.ImageTag = imageTag
+				}
+				ctags := GetOTelContainerTags(res.Attributes())
+				for _, ctag := range ctags {
+					if v, ok := c.containerTagCache.Get(cid); ok {
+						cacheVal := v.(*sync.Map)
+						// check if the key already exists in the cache
+						if _, ok := cacheVal.Load(ctag); ok {
+							continue
+						}
+						cacheVal.Store(ctag, struct{}{})
+					} else {
+						cacheVal := &sync.Map{}
+						cacheVal.Store(ctag, struct{}{})
+						c.containerTagCache.Set(cid, cacheVal, cache.DefaultExpiration)
+					}
+				}
```

Comment:
Is there a way to void this cache, like concentrator taking in ctags etc ?

Reply:
We can just store container tags in a plain map instead, though that means the container tags can grow unbounded. I don't know how much is that a problem in practice, WDYT?

On the other hand concentrator processes and flushes data asynchronously, it means there will be a time gap between when container tags are stored vs. when they are retrieved. We must ensure the container tags are still present by the time they are retrieved.

Reply:
No. i meant why can't concentrator take in tags just like containerID, as it works on computing stats ? Why do we need to do the look up post aggregation .